imsim/photon_ops.py

```diff
@@ -135,12 +119,248 @@ def applyTo(self, photon_array, local_wcs=None, rng=None):
+        The telescope to trace through.
+    latitude : float
+        Geographic latitude of telescope. Needed to calculate the field rotation
+        for the spider diffraction.
+    altitude, azimuth : float
+        alt/az coordinates the telescope is pointing to (in degree).
+    sky_pos : galsim.CelestialCoord
+    icrf_to_field : galsim.GSFitsWCS
+    shift_optics : dict[str, list[float]]
+        A dict mapping optics keys to shifts represented by a list of 3 floats.
+        The corresponding optics will be displaced by the specified corrdinates.
+        Example config for perturbed+defocused telescope to obtain a donut:
+        -
+            type: lsst_optics
+            ...
+            shift_optics:
+              Detector: [0, 0, 1.5e-3]
+              M2: [3.0e-3, 0, 0]
+    seed : An optional seed forwarded to the angle distribution used
+           for diffraction.
+    """
+
+    _req_params = {
+        "telescope": Optic,
+        "band": str,
+        "latitude": float,
+    }
+    _opt_params = {"shift_optics": dict}
+
+    def __init__(
+        self,
+        telescope,
+        latitude,
+        altitude,
+        azimuth,
+        sky_pos,
+        icrf_to_field,
+        shift_optics=None,
+        seed=None,
+    ):
+        if shift_optics is not None:
+            for optics_key, shift in shift_optics.items():
+                telescope = telescope.withGloballyShiftedOptic(optics_key, shift)
+        self.telescope = telescope
+        self.latitude = latitude
+        self.altitude = altitude
+        self.azimuth = azimuth
+        self.sky_pos = sky_pos
+        self.icrf_to_field = icrf_to_field
+
+        deviate = GaussianDeviate(seed=seed)
+
+        def diffraction_rng(phi):
+            var = phi**2
+            deviate.generate_from_variance(var)
+            return var
+
+        self.diffraction_rng = diffraction_rng
+
+    def applyTo(self, photon_array, local_wcs=None, rng=None):
+        """Apply the photon operator to a PhotonArray.
+
+        Here, we assume that the photon array has passed through
+        `imsim.atmPSF.AtmosphericPSF` which stores sampled pupil
+        locations in photon_array.pupil_u and .pupil_v.
+
+        Parameters
+        ----------
+        photon_array:   A `PhotonArray` to apply the operator to.
+        local_wcs:      A `LocalWCS` instance defining the local WCS for the current photon
+                        bundle in case the operator needs this information.  [default: None]
+        rng:            A random number generator to use if needed. [default: None]
+        """
+
+        xy_to_v = XyToV(local_wcs, self.icrf_to_field, self.sky_pos)
+        # Convert xy coordinates to a cartesian 3d velocity vector of the photons
+        v = xy_to_v(photon_array.x, photon_array.y)
+
+        # Adjust for refractive index of air
+        wavelength = photon_array.wavelength * 1e-9
+        n = self.telescope.inMedium.getN(wavelength)
+        v /= n[:, None]
```

Comment:
Are you sure this line is right?  Why does the refractive index of air affect the diffraction angle?

Reply:
This is coming from @jmeyers314 . Do you have more info about this?

Reply:
It's photon velocity in units of `c`, so the refractive index does need to be accounted for.